demo_apps/whatsapp_llama2.md

```diff
@@ -0,0 +1,151 @@
+
+## Writing Llama 2 Enabled Web App
+
+The Python-based web app we developed uses [LangChain](https://www.langchain.com/), an open source LLM development framework, and [Replicate](https://replicate.com/), a service provider hosting LLM models in the cloud, to receive the user query sent by the webhook, which will be covered in the next section, pass the query to Llama 2, and send the Llama 2 answer back to the webhook. For more information on how to use LangChain and Replicate to develop Llama 2 apps, see the [Llama 2 demo apps](https://github.com/facebookresearch/llama-recipes/tree/main/demo_apps) - you may also replace Replicate with other Llama deployment options shown in the demo apps.
+
+First, let's create a new conda (or you can use venv if you like) environment and install all the required packages:
+
+```
+conda create -n whatsapp-llama python=3.8
+conda activate whatsapp-llama
+pip install langchain replicate flask requests uvicorn gunicorn
+```
+
+Then, create a Python file named llama-chatbot.py with the following code, which defines a class `WhatsAppClient` and a method `send_text_message` to post a message (the answer generated by Llama 2 on a user query) to the WhatsApp Cloud API, which then sends the answer back to the WhatsApp user. Remember to set `WHATSAPP_API_TOKEN` and `WHATSAPP_CLOUD_NUMBER_ID` to the values you saved in the previous section.
+
+```
+import langchain
+from langchain.llms import Replicate
+from flask import Flask
+from flask import request
+import os
+import requests
+import json
+
+class WhatsAppClient:
+    API_URL = "https://graph.facebook.com/v17.0/"
+    WHATSAPP_API_TOKEN = "<Temporary access token from your WhatsApp API Setup>"
+    WHATSAPP_CLOUD_NUMBER_ID = "<Phone number ID from your WhatsApp API Setup>"
+
+    def __init__(self):
+        self.headers = {
+            "Authorization": f"Bearer {self.WHATSAPP_API_TOKEN}",
+            "Content-Type": "application/json",
+        }
+        self.API_URL = self.API_URL + self.WHATSAPP_CLOUD_NUMBER_ID
+        
+    def send_text_message(self, message, phone_number):
+        payload = {
+            "messaging_product": 'whatsapp',
+            "to": phone_number,
+            "type": "text",
+            "text": {
+                "preview_url": False,
+                "body": message
+            }
+        }
+        response = requests.post(f"{self.API_URL}/messages", json=payload, headers=self.headers)
+        return response.status_code
+```
+
+Finally, add the code below to llama-chatbot.py, which creates a Llama 2 instance and defines an HTTP method `msgrcvd` to:
+1. receive the user message forwarded by the webhook;
+2. ask Llama 2 for the answer;
+3. call the `WhatsAppClient`'s `send_text_message`` with a recipient's phone number.
+   
+```   
+os.environ["REPLICATE_API_TOKEN"] = "<your replicate api token>"    
```

Comment:
Same here, we can point out that a replicate account is needed, as obvious as it is.

Reply:
I mentioned Replicate and other options in the first paragraph under "Writing Llama 2 Enabled Web App" so think that should be good.